dali/imgcodec/decoders/tiff_libtiff.cc

```diff
@@ -0,0 +1,228 @@
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_ROWSPERSTRIP, &info.rows_per_strip));
+
+  info.is_tiled = TIFFIsTiled(tiffptr);
+
+  uint16_t photometric_interpretation;
+  if (TIFFGetField(tiffptr, TIFFTAG_PHOTOMETRIC, &photometric_interpretation) &&
+      photometric_interpretation == PHOTOMETRIC_PALETTE) {
+    info.is_palette = true;
+  }
+
+  return info;
+}
+
+}  // namespace detail
+
+DecodeResult LibTiffDecoderInstance::Decode(SampleView<CPUBackend> out, ImageSource *in,
+                                           DecodeParams opts, const ROI &requested_roi) {
+  auto tiff = detail::OpenTiff(in);
+  auto info = detail::GetTiffInfo(tiff.get());
+
+  // TODO(skarpinski) Support other bit depths
+  DALI_ENFORCE(info.bit_depth == 8, "Not implemented: Only bit depth = 8 is supported");
+
+  // TODO(skarpinski) Support palette images
+  DALI_ENFORCE(!info.is_palette, "Not implemented: Palette images are not yet supported");
+
+  // TODO(skarpinski) Support tiled images
+  DALI_ENFORCE(!info.is_tiled, "Not implemented: Tiled images are not yet supported");
+
+  unsigned out_channels = NumberOfChannels(opts.format, info.channels);
+
+  // TODO(skarpinski) support different types
+  using InType = uint8_t;
+  using OutType = uint8_t;
+
+  auto row_nbytes = TIFFScanlineSize(tiff.get());
+  std::unique_ptr<InType, void(*)(void*)> row_buf{
+    static_cast<InType *>(_TIFFmalloc(row_nbytes)), _TIFFfree};
+  DALI_ENFORCE(row_buf.get() != nullptr, "Could not allocate memory");
+  InType * const row_in  = row_buf.get();
+  OutType * const img_out = out.mutable_data<OutType>();
+
+  ROI roi;
+  if (!requested_roi.use_roi()) {
+    roi.begin = {0, 0};
+    roi.end = out.shape().first(2);
+  } else {
+    roi = requested_roi;
+  }
+
+  // Need to read sequentially since not all the images support random access
+  // From: http://www.libtiff.org/man/TIFFReadScanline.3t.html
+  // Compression algorithm does not support random access. Data was requested in a non-sequential
+  // order from a file that uses a compression algorithm and that has RowsPerStrip greater than one.
+  // That is, data in the image is stored in a compressed form, and with multiple rows packed into a
+  // strip. In this case, the library does not support random access to the data. The data should
+  // either be accessed sequentially, or the file should be converted so that each strip is made up
+  // of one row of data.
+  const bool allow_random_row_access = (info.compression == COMPRESSION_NONE
+                                        || info.rows_per_strip == 1);
+  if (!allow_random_row_access) {
+    for (int64_t y = 0; y < roi.begin[0]; y++) {
+      LIBTIFF_CALL(TIFFReadScanline(tiff.get(), row_in, y, 0));
+    }
+  }
+
+  uint64_t out_row_stride = roi.shape()[1] * out_channels;
+
+  DALIImageType in_format;
+  if (info.channels == 1)
+    in_format = DALI_GRAY;
+  else if (info.channels >= 3)
+    in_format = DALI_RGB;
+  else
+    in_format = DALI_ANY_DATA;
```

Comment:
For DALI, I think we should mark as RGB only if channels == 3

Reply:
You're right, treating everything as RGB is wrong, but enforcing channels == 3 would make us unable to decode RGBA for example. I'd propose to do the following (and this would make us compatible with [the old decoder](https://github.com/NVIDIA/DALI/blob/d5b543bf77a01be62296c032652a839c9b376192/dali/image/tiff_libtiff.cc#L155-L188):
```cpp
  if (info.channels == 1)
    in_format = DALI_GRAY;
  else if (info.channels >= 3 && opts.format != DALI_ANY_DATA)
    in_format = DALI_RGB;
  else
    in_format = DALI_ANY_DATA;
```
In other words: if we have a multi-channel TIFF and the user requested output in format other that `DALI_ANY_DATA`, we will treat the data as RGB, i.e. ignore the extra channels and convert. But if the user requested `DALI_ANY_DATA`, then we treat the TIFF as `DALI_ANY_DATA`. 

Is that fine?

Reply:
As we discussed offline, I changed this part as described above